hoot-core/src/main/cpp/hoot/core/algorithms/alpha-shape/AlphaShapeGenerator.cpp

```diff
@@ -106,76 +110,111 @@ std::shared_ptr<Geometry> AlphaShapeGenerator::generateGeometry(OsmMapPtr inputM
-  if (!cutterShape)
+  if (!_geometry)
   {
-    cutterShape.reset(GeometryFactory::getDefaultInstance()->createEmptyGeometry());
+    _geometry.reset(GeometryFactory::getDefaultInstance()->createEmptyGeometry());
   }
-  cutterShape.reset(cutterShape->buffer(_buffer));
+  _geometry.reset(_geometry->buffer(_buffer));
 
   // See _coverStragglers description. This is an add-on behavior that is separate from the original
   // Alpha Shape algorithm.
   if (_manuallyCoverSmallPointClusters)
   {
-    _coverStragglers(cutterShape, inputMap);
+    _coverStragglers(inputMap);
   }
 
-  return cutterShape;
+  return _geometry;
 }
 
-void AlphaShapeGenerator::_coverStragglers(std::shared_ptr<Geometry>& geometry,
-                                           const ConstOsmMapPtr& map)
+void AlphaShapeGenerator::_coverStragglers(const ConstOsmMapPtr& map)
 {
   LOG_DEBUG("Covering stragglers...");
 
   // Pretty simple...go through and find any point that wasn't covered by the Alpha Shape and draw
   // a buffer around it. This can get very slow for fairly large linear datasets. May need a new alg
   // here, if possible, for better performance.
   const NodeMap& nodes = map->getNodes();
-  vector<GeometryPtr> stragglers;
-  int processed = 0;
+  //  Push all nodes onto the work queue
   for (NodeMap::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
+    _nodes.push(it->second);
+  //  Start all of the threads
+  std::vector<std::thread> threads;
+  for (int i = 0; i < _maxThreads; ++i)
+    threads.push_back(thread(&AlphaShapeGenerator::_coverStragglersWorker, this));
+  //  Set working to false after starting all threads so that they end once the queue is empty
+  _working = false;
+  //  Wait on all threads to end
+  for (int i = 0; i < _maxThreads; ++i)
+    threads[i].join();
+
+  if (_stragglers.size() > 0)
   {
-    NodePtr node = it->second;
-    GeometryPtr point(
-      GeometryFactory::getDefaultInstance()->createPoint(
-        geos::geom::Coordinate(node->getX(), node->getY())));
-    if (!geometry->contains(point.get()))
-    {
-      LOG_TRACE(
-        "Point " << point->toString() << " not covered by alpha shape. Buffering and adding it...");
-      point.reset(point->buffer(_buffer));
-      stragglers.push_back(point);
-    }
-    processed++;
+    LOG_DEBUG("Adding " << StringUtils::formatLargeNumber(_stragglers.size())
+              << " point stragglers to alpha shape.");
+    //  Merge the stragglers geometries
+    GeometryPtr merged = GeometryMerger().mergeGeometries(_stragglers, *_geometry->getEnvelopeInternal());
+    //  Join the original geometry with the stragglers geometry
+    _geometry.reset(_geometry->Union(merged.get()));
   }
+  LOG_VART(_geometry->getArea());
+}
 
-  LOG_VARD(processed);
-
-
-  if (stragglers.size() > 0)
+void AlphaShapeGenerator::_coverStragglersWorker()
+{
+  //  Make a thread local copy of the geometry because Geometry::contains() is not thread-safe
+  GeometryPtr copy(_geometry->clone());
```

Comment:
This is important, it is a small memory overhead but allows us to call `Geometry::contains()` in the threads because it isn't thread safe so one copy per thread overcomes that.